### PR TITLE
Construct Transaction API

### PR DESCRIPTION
**Pull request by n8maninger**
Adds two new endpoints to construct transactions. This combines and simplifies the existing fund flow for sending siacoin and siafund transactions.

Requires #210

### CODE REVIEW

**Comment by lukechampine** [comment]
Definitely not safe to blindly sign whatever hashes a server gives you. 😬 This is a good illustration of how, strictly speaking, it's not *possession* of private keys that matters, but *control*.

Clients need to know what they're signing, which means they need to compute the hash. And they can't compute the hash blindly either (i.e. call BLAKE2b on a buffer returned by the server) for the reason. At best, the server could return a buffer with preallocated space for the client to write the SiacoinOutputs into. But that's still vulnerable: the server could trick the client by having it write the outputs into `ArbitraryData`, while the actual SiacoinOutputs are malicious. So the client needs to be able to parse transactions. As far as I can tell, this is unavoidable.

IMO, any client should have the ability to parse transactions anyway (so that it can display them to the user), in addition to managing private keys and hashing/signing.

**Comment by n8maninger** [comment]
> Definitely not safe to blindly sign whatever hashes a server gives you. 😬 This is a good illustration of how, strictly speaking, it's not possession of private keys that matters, but control.

You’re absolutely right that blindly signing hashes from an untrusted server is a security risk. However, this API is intended for trusted scenarios where the user controls both the client and the server.

> IMO, any client should have the ability to parse transactions anyway (so that it can display them to the user), in addition to managing private keys and hashing/signing.

Requiring an implementation of full transaction parsing and sighash calculation in every possible language is really poor devex. Displaying a JSON object in a UI is relatively straightforward, implementing full sighash calculation is much more complex. For most languages, ed25519 libraries are available, but Sia specific transaction hashing logic isn’t.

The goal is to provide a minimal, generic, and language-agnostic flow for developers to interact with that balances usability and security for trusted implementations. For users that need maximum security and control, they still have the existing fund flow and can calculate the sig hash manually in Go or Rust. 

I'm sure we can come up with a flow that is *secure enough* for daily use without requiring developers to dig into low-level Sia specific code. Maybe it's as simple as separating it a bit so the user has to pass the transaction back to the server and can inspect it rather than caching it server-side?

**Comment by n8maninger** [comment]
@lukechampine I removed the signing flow from this. Now it just builds a transaction and returns it for signing.

File: api/server.go [hunk 1]
```diff
@@ -674,6 +676,376 @@ func (s *server) walletsFundSFHandler(jc jape.Context) {
...
+	var sent bool
+	var locked []types.Hash256
+	defer func() {
+		if sent {
+			return
+		}
+		s.wm.Release(locked)
+	}()
+
+	sces, basis, siacoinChange, err := s.wm.SelectSiacoinElements(walletID, siacoinInput.Add(fee), false)
```
Comment (peterjan) [inline api/server.go:740]: `useUnconfirmed` is probably something we want to expose in the request object?

File: .changeset/add_transaction_construction_api.md
Comment (ChrisSchinnerl) [inline .changeset/add_transaction_construction_api.md:1]: This appears to be outdated now

File: api/server.go [hunk 2]
```diff
@@ -674,6 +676,376 @@ func (s *server) walletsFundSFHandler(jc jape.Context) {
...
+		})
+	}
+
+	knownAddresses := make(map[types.Address]wallet.Address)
+	getAddress := func(addr types.Address) (wallet.Address, error) {
+		if a, ok := knownAddresses[addr]; ok {
+			return a, nil
+		}
+		a, err := s.wm.WalletAddress(walletID, addr)
+
```
Comment (ChrisSchinnerl) [inline api/server.go:778]: Both handlers got an unnecessary newline here.

File: api/server.go [hunk 3]
```diff
@@ -674,6 +676,376 @@ func (s *server) walletsFundSFHandler(jc jape.Context) {
...
+	if err := jc.DecodeParam("id", &walletID); err != nil {
+		return
+	}
+	var wcr WalletConstructRequest
+	if err := jc.Decode(&wcr); err != nil {
+		return
+	}
+
+	if wcr.ChangeAddress == types.VoidAddress {
+		jc.Error(errors.New("change address must be specified"), http.StatusBadRequest)
```
Comment (ChrisSchinnerl) [inline api/server.go:695]: Any reason why v1 checks for a change address immediately but v2 only does it when there actually is change?

File: api/server.go [hunk 4]
```diff
@@ -674,6 +676,376 @@ func (s *server) walletsFundSFHandler(jc jape.Context) {
...
+		SiacoinInputs:  make([]types.SiacoinInput, 0, len(sces)),
+		SiacoinOutputs: wcr.Siacoins,
+		SiafundInputs:  make([]types.SiafundInput, 0, len(sfes)),
+		SiafundOutputs: wcr.Siafunds,
+	}
+
+	for _, sce := range sces {
+		addr, err := getAddress(sce.SiacoinOutput.Address)
+		if err != nil {
+			jc.Error(fmt.Errorf("failed to get address: %w", err), http.StatusInternalServerError)
```
Comment (peterjan) [inline api/server.go:802]: want to handle `wallet.ErrNotFound` and return a 404? If we do we might as well make it part of the validation up top

File: api/server.go [hunk 5]
```diff
@@ -674,6 +676,376 @@ func (s *server) walletsFundSFHandler(jc jape.Context) {
...
+		txn.SiacoinInputs = append(txn.SiacoinInputs, sci)
+		txn.Signatures = append(txn.Signatures, types.TransactionSignature{
+			ParentID: types.Hash256(sce.ID),
+			CoveredFields: types.CoveredFields{
+				WholeTransaction: true,
+			},
+		})
+	}
+
+	for _, sfe := range sfes {
```
Comment (ChrisSchinnerl) [inline api/server.go:829]: Not that it matters but bringing it up for consistency. One handler does siacoins first and then siafunds and the other one does siafunds first and then siacoins

File: .changeset/add_transaction_construction_api.md [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+default: minor
+---
+
+# Add transaction construction API
+
+Added two new endpoints to construct transactions with a few restrictions for ease of use. Clients can still use the existing fund endpoints for "advanced" transactions. All addresses in the wallet must all have either unlock conditions with a single required signature or a public key spend policy.
+
+This is a two step process. The private keys are never transmitted to the server.
+
+The client first calls `[POST] /api/:wallet/transaction/construct` with the recipients. The server will construct the transaction and return a list of hashes that the client needs to sign to broadcast the transaction. The client needs to match the returned public keys to their ed25519 private key and sign all of the hashes.
```
Comment (peterjan) [inline .changeset/add_transaction_construction_api.md:11]: the routes are wrong here, should be `/api/wallts/:id/transaction/construct`

File: api/server.go [hunk 7]
```diff
@@ -674,6 +676,376 @@ func (s *server) walletsFundSFHandler(jc jape.Context) {
...
+		addr, err := getAddress(sce.SiacoinOutput.Address)
+		if err != nil {
+			jc.Error(fmt.Errorf("failed to get address: %w", err), http.StatusInternalServerError)
+			return
+		}
+
+		sci := types.V2SiacoinInput{
+			Parent: sce,
+			SatisfiedPolicy: types.SatisfiedPolicy{
+				Policy: *addr.SpendPolicy,
```
Comment (ChrisSchinnerl) [inline api/server.go:1029]: You are setting the policy here but then check if it is not nil and set it again

File: api/server.go [hunk 8]
```diff
@@ -674,6 +676,376 @@ func (s *server) walletsFundSFHandler(jc jape.Context) {
...
+	for _, sfe := range sfes {
+		addr, err := getAddress(sfe.SiafundOutput.Address)
+		if err != nil {
+			jc.Error(fmt.Errorf("failed to get address: %w", err), http.StatusInternalServerError)
+			return
+		}
+
+		sfi := types.V2SiafundInput{
+			Parent:       sfe,
+			ClaimAddress: wcr.ChangeAddress,
```
Comment (ChrisSchinnerl) [inline api/server.go:998]: I think we might hit this without checking whether `ChangeAddress` is set. Potentially allowing for a claim address that is the void address.

File: api/client.go [hunk 9]
```diff
@@ -334,6 +338,26 @@ func (c *WalletClient) FundSF(txn types.Transaction, amount uint64, changeAddr,
 	return
 }
 
+// Construct constructs a transaction and returns its ID
```
Comment (peterjan) [inline api/client.go:341]: outdated docstring

File: api/client.go [hunk 10]
```diff
@@ -334,6 +338,26 @@ func (c *WalletClient) FundSF(txn types.Transaction, amount uint64, changeAddr,
...
+func (c *WalletClient) Construct(siacoins []types.SiacoinOutput, siafunds []types.SiafundOutput, change types.Address) (resp WalletConstructResponse, err error) {
+	err = c.c.POST(fmt.Sprintf("/wallets/%v/construct/transaction", c.id), WalletConstructRequest{
+		Siacoins:      siacoins,
+		Siafunds:      siafunds,
+		ChangeAddress: change,
+	}, &resp)
+	return
+}
+
+// ConstructV2 constructs a v2 transaction and returns its ID
```
Comment (peterjan) [inline api/client.go:351]: outdated docstring

File: .changeset/add_transaction_construction_api.md [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+
+After signing, the client calls `[POST] /api/:wallet/transaction/construct/:id` with the array of signatures. The server will add the signatures to the transaction and broadcast it. If the client provided the correct signatures, the transaction will be added to the tpool and broadcast.
```
Comment (peterjan) [inline .changeset/add_transaction_construction_api.md:13]: this paragraph is not what's happening right? I'm assuming this is a somewhat outdated version that didn't get updated after refactoring the user flow

File: api/api_test.go
Comment (ChrisSchinnerl) [inline api/api_test.go:1]: Can you extend the tests with edge cases around the change address?

File: api/server.go [hunk 12]
```diff
@@ -273,6 +275,7 @@ func (s *server) txpoolParentsHandler(jc jape.Context) {
 
 func (s *server) txpoolTransactionsHandler(jc jape.Context) {
 	jc.Encode(TxpoolTransactionsResponse{
+		Basis:          s.cm.Tip(),
```
Comment (peterjan) [inline api/server.go:278]: Do we still want to add the basis to the response here? I checked and we're never using it, so I'm assuming it may have been added at one point for testing, then refactored and overlooked.

File: api/server.go [hunk 13]
```diff
@@ -674,6 +676,376 @@ func (s *server) walletsFundSFHandler(jc jape.Context) {
...
+	for _, sfe := range sfes {
+		addr, err := getAddress(sfe.SiafundOutput.Address)
+		if err != nil {
+			jc.Error(fmt.Errorf("failed to get address: %w", err), http.StatusInternalServerError)
+			return
+		}
+
+		sfi := types.SiafundInput{
+			ParentID:     sfe.ID,
+			ClaimAddress: wcr.ChangeAddress,
```
Comment (ChrisSchinnerl) [inline api/server.go:838]: Maybe I don't know enough about how the claim address works but it seems a bit odd to me to implicitly use the change address as the claim address. 
e.g. person A has 2 siafunds. Sends 1 siafund to person B and 1 siafund back to itself via the change address. Does that mean person A still receives the siacoins the siafund generates which they just sent to person B?

File: api/api_test.go [hunk 14]
```diff
@@ -1286,6 +1286,612 @@ func TestConsensusUpdates(t *testing.T) {
 	}
 }
 
+func TestConstructSiacoins(t *testing.T) {
```
Comment (peterjan) [inline api/api_test.go:1289]: these tests are great but it's really only the happy path, might be worth throwing a couple extra lines in to test whether we properly release locked outputs

File: api/server.go (hunk 8 above)
Comment (n8maninger) [reply api/server.go:998]: That's why it was changed to a global check in the v1 handler

File: api/api_test.go
Comment (n8maninger) [reply api/api_test.go:1]: What kind of edge cases are you thinking of? The test already checks that it matches the expected address in the test. It's either set correctly or not set correctly.

File: api/server.go (hunk 5 above)
Comment (n8maninger) [reply api/server.go:829]: `UpdateV2TransactionSet` requires a V2Transaction as an argument, since we want to update the Siafund proofs to match the Siacoin proofs, we add the Siafund elements to the transaction first then check the basis and update the Siafund proofs.

Comment (n8maninger) [reply api/server.go:829]: UpdateV2TransactionSet requires a V2Transaction as an argument, since we want to update the Siafund proofs to match the Siacoin proofs, we add the Siafund elements to the transaction first then check the basis and update the Siafund proofs.

File: api/server.go (hunk 1 above)
Comment (n8maninger) [reply api/server.go:740]: I don't want to expose it here to keep it simple

File: api/server.go (hunk 12 above)
Comment (n8maninger) [reply api/server.go:278]: Technically no for V1, but `Client.TxpoolBroadcast` has the basis as its first argument. Good to be consistent and provide the user a value to put there.

Comment (n8maninger) [reply api/server.go:278]: Absolutely. The basis is very important for broadcasting v2 transactions.

File: api/server.go (hunk 5 above)
Comment (n8maninger) [reply api/server.go:837]: There's no technical reason we couldn't do it the other way around, just felt weird.

Comment (n8maninger) [reply api/server.go:837]: Added a comment on *why* the ordering is the way it is:
```go
// the siafund elements are added to the transaction first because `UpdateV2TransactionSet` takes
// a V2 transaction as an argument. The Siacoin basis is our target because the transaction is
// guaranteed to have a non-zero Siacoin basis while the Siafund basis will be zero when not 
// sending Siafunds.
```

**Review by peterjan — DISMISSED** [review]
LGTM 👍

**Review by ChrisSchinnerl — DISMISSED** [review]
